File: frontend-vue/src/api/memory.ts
import api from './index'

// ── 长期记忆（记忆四层架构 L2 档案卡：跨会话留存） ──

export type MemorySlot = 'identity' | 'preference' | 'decision' | 'fact'
export type MemorySource = 'user_confirmed' | 'derived' | 'tool_written'

export interface MemoryEntry {
  id: string
  slot: string
  slot_label: string
  key: string
  value: string
  confidence: number
  source: string
  source_label: string
  has_embedding: boolean
  access_count: number
  last_accessed_at: string | null
  status: string
  created_at: string
  updated_at: string
}

export interface MemorySearchHit extends MemoryEntry {
  similarity: number
  score: number
}

export interface OrganizeResult {
  backfilled: number
  merged: number
  archived: number
  evicted: number
  errors: string[]
}

export interface OrganizeStatus {
  running: boolean
  started_at: number | null
  finished_at: number | null
  result: OrganizeResult | null
  error: string | null
}

export interface ProfileListResponse {
  entries: MemoryEntry[]
  counts: Record<string, number>
  total: number
  slots: { slot: string; label: string }[]
  organize: OrganizeStatus
}

/** 整卡列表（按槽位分组统计） */
export async function listMemory(includeArchived = false): Promise<ProfileListResponse> {
  const { data } = await api.get('/v1/memory/profile', {
    params: includeArchived ? { archived: 'true' } : {},
  })
  return data
}

/** 新建 / 更新一条记忆（同 slot+key 自动更新） */
export async function upsertMemory(body: {
  slot: MemorySlot
  key: string
  value: string
  confidence?: number
  source?: MemorySource
}): Promise<{ entry: MemoryEntry; created: boolean; duplicate_of?: MemoryEntry; evicted?: number }> {
  const { data } = await api.post('/v1/memory/profile', body)
  return data
}

/** 编辑已有记忆（按 id） */
export async function updateMemory(body: {
  id: string
  key?: string
  value?: string
  confidence?: number
  source?: MemorySource
}): Promise<{ entry: MemoryEntry }> {
  const { data } = await api.put('/v1/memory/profile', body)
  return data
}

/** 删除单条记忆 */
export async function deleteMemory(id: string): Promise<void> {
  await api.delete(`/v1/memory/profile/${encodeURIComponent(id)}`)
}

/** 清空全部记忆（需 confirm=true） */
export async function clearMemory(): Promise<{ deleted: number }> {
  const { data } = await api.post('/v1/memory/profile/clear', { confirm: true })
  return data
}

/** 语义检索（相似度 + 重排序；embedding 不可用时降级关键词） */
export async function searchMemory(
  query: string,
  opts?: { top_k?: number; slot?: MemorySlot },
): Promise<{ query: string; mode: 'semantic' | 'keyword'; count: number; results: MemorySearchHit[] }> {
  const { data } = await api.post('/v1/memory/search', {
    query,
    top_k: opts?.top_k ?? 10,
    slot: opts?.slot,
  })
  return data
}

/** 触发异步智能整理（去重 / 归档 / 补嵌入） */
export async function startOrganize(): Promise<{ started: boolean; status: OrganizeStatus }> {
  const { data } = await api.post('/v1/memory/organize')
  return data
}

/** 整理任务状态 */
export async function getOrganizeStatus(): Promise<OrganizeStatus> {
  const { data } = await api.get('/v1/memory/organize/status')
  return data?.status
}

export const MEMORY_SLOTS: { slot: MemorySlot; label: string }[] = [
  { slot: 'identity', label: '身份' },
  { slot: 'preference', label: '偏好' },
  { slot: 'decision', label: '关键决策' },
  { slot: 'fact', label: '长期事实' },
]

// ── 冲突裁决 API ──

export interface MemoryConflict {
  conflict_id: string
  slot: string
  item_key: string
  old_value: string
  new_value: string
  source: string
  created_at: number
}

export interface ConflictListResponse {
  conflicts: MemoryConflict[]
  count: number
}

/** 获取待裁决冲突列表 */
export async function listConflicts(): Promise<ConflictListResponse> {
  const { data } = await api.get('/v1/memory/conflicts')
  return data
}

export interface ResolveConflictResponse {
  conflict_id: string
  final_value: string
  resolution: string
  slot: string
  item_key: string
  profile_update?: {
    success: boolean
    item: string | null
  }
}

/**
 * 裁决冲突
 * @param conflictId 冲突 ID
 * @param resolution 裁决方式: 'keep_old' | 'use_new' | 'manual'
 * @param manualValue 手动修改值（仅当 resolution === 'manual'）
 */
export async function resolveConflict(
  conflictId: string,
  resolution: 'keep_old' | 'use_new' | 'manual',
  manualValue?: string,
): Promise<ResolveConflictResponse> {
  const body: { resolution: string; manual_value?: string } = { resolution }
  if (resolution === 'manual' && manualValue !== undefined) {
    body.manual_value = manualValue
  }
  const { data } = await api.post(`/v1/memory/conflicts/${encodeURIComponent(conflictId)}/resolve`, body)
  return data.conflict
}

/** 删除冲突（用户否认） */
export async function deleteConflict(conflictId: string): Promise<{ deleted: string }> {
  const { data } = await api.delete(`/v1/memory/conflicts/${encodeURIComponent(conflictId)}`)
  return data
}

File: python-engine/app/api/memory.py
"""记忆管理 API — L2 档案卡（用户长期记忆）。

路由（经 Go 网关 newProxy 代理，自动追加 ?user_id=<claims.UserID>）：
- GET    /v1/memory/profile                 整卡列表（按槽位分组统计）
- POST   /v1/memory/profile                 新建/更新条目 {slot, key, value, confidence?, source?}
- PUT    /v1/memory/profile                 编辑条目 {id, key?, value?, confidence?, source?}
- DELETE /v1/memory/profile/{id}            删除单条
- POST   /v1/memory/profile/clear           清空全部记忆 {confirm: true}
- POST   /v1/memory/search                  语义检索（相似度 + 重排序）
- POST   /v1/memory/organize                触发异步智能整理（去重/归档/补嵌入）
- GET    /v1/memory/organize/status         整理任务状态
"""
from __future__ import annotations

import logging

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from app.memory.layers import SLOTS, SLOT_LABELS
from app.memory.service import get_service

logger = logging.getLogger(__name__)

router = APIRouter(tags=["memory"])


def _scope(request: Request) -> tuple[str, str]:
    """从网关注入的 query 参数解析 (tenant_id, user_id)。"""
    user_id = (request.query_params.get("user_id") or "").strip()
    tenant_id = (request.query_params.get("tenant_id") or "default").strip()
    return tenant_id, user_id


def _unavailable() -> JSONResponse:
    return JSONResponse(
        {"success": False, "error": "memory service unavailable (PostgreSQL required)"},
        status_code=503,
    )


def _bad_request(msg: str) -> JSONResponse:
    return JSONResponse({"success": False, "error": msg}, status_code=400)


@router.get("/v1/memory/profile")
async def list_profile(request: Request):
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    include_archived = request.query_params.get("archived") == "true"
    data = await svc.list_entries(tenant_id, user_id, include_archived=include_archived)
    return {
        "success": True,
        **data,
        "slots": [{"slot": s, "label": SLOT_LABELS[s]} for s in SLOTS],
        "organize": svc.organize_status(tenant_id, user_id),
    }


@router.post("/v1/memory/profile")
async def upsert_profile(request: Request):
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    body = await request.json()
    try:
        confidence = int(body.get("confidence", 50))
        confidence = max(0, min(100, confidence))
        result = await svc.upsert(
            tenant_id=tenant_id,
            user_id=user_id,
            slot=str(body.get("slot") or "fact"),
            key=str(body.get("key") or ""),
            value=str(body.get("value") or ""),
            confidence=confidence,
            source=str(body.get("source") or "user_confirmed"),
        )
    except ValueError as e:
        return _bad_request(str(e))
    return {"success": True, **result}


@router.put("/v1/memory/profile")
async def update_profile(request: Request):
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    body = await request.json()
    entry_id = str(body.get("id") or "")
    if not entry_id:
        return _bad_request("id is required")
    try:
        confidence = body.get("confidence")
        if confidence is not None:
            confidence = max(0, min(100, int(confidence)))
        entry = await svc.update_entry(
            tenant_id, user_id, entry_id,
            key=str(body["key"]) if body.get("key") is not None else None,
            value=str(body["value"]) if body.get("value") is not None else None,
            confidence=int(confidence) if confidence is not None else None,
            source=str(body["source"]) if body.get("source") else None,
        )
    except ValueError as e:
        return _bad_request(str(e))
    if entry is None:
        return JSONResponse({"success": False, "error": "entry not found"}, status_code=404)
    return {"success": True, "entry": entry.to_dict()}


@router.delete("/v1/memory/profile/{entry_id}")
async def delete_profile(entry_id: str, request: Request):
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    deleted = await svc.delete_entry(tenant_id, user_id, entry_id)
    if not deleted:
        return JSONResponse({"success": False, "error": "entry not found"}, status_code=404)
    return {"success": True, "deleted": entry_id}


@router.post("/v1/memory/profile/clear")
async def clear_profile(request: Request):
    """清空当前用户全部记忆（隐私出口）。要求 body {confirm: true} 防误触。"""
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    body = await request.json()
    if not body.get("confirm"):
        return _bad_request("confirm=true is required to clear all memories")
    count = await svc.clear_all(tenant_id, user_id)
    return {"success": True, "deleted": count}


@router.post("/v1/memory/search")
async def search_memory(request: Request):
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    body = await request.json()
    try:
        data = await svc.search(
            tenant_id, user_id,
            query=str(body.get("query") or ""),
            top_k=int(body.get("top_k", 10)),
            slot=str(body["slot"]) if body.get("slot") else None,
        )
    except ValueError as e:
        return _bad_request(str(e))
    return {"success": True, **data}


@router.post("/v1/memory/organize")
async def organize_memory(request: Request):
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    result = await svc.start_organize(tenant_id, user_id)
    return {"success": True, **result, "status": svc.organize_status(tenant_id, user_id)}


@router.get("/v1/memory/organize/status")
async def organize_status(request: Request):
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    return {"success": True, "status": svc.organize_status(tenant_id, user_id)}


# ── L3 摘要管理 ──────────────────────────────────────

@router.get("/v1/memory/summaries")
async def list_summaries(request: Request):
    """列出摘要记忆（管理端审计）。"""
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    limit = int(request.query_params.get("limit", "50") or 50)
    # S 修复：限制上限，避免非法/超大 limit 导致 DB 无界查询
    limit = min(max(limit, 1), 200)
    data = await svc.list_summaries(tenant_id, user_id, limit)
    return {"success": True, **data}


# ── 冲突裁决 ─────────────────────────────────────────

@router.get("/v1/memory/conflicts")
async def list_conflicts(request: Request):
    """列出待裁决的记忆冲突。"""
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    conflicts = await svc.list_conflicts(tenant_id, user_id)
    return {"success": True, "conflicts": conflicts, "count": len(conflicts)}


@router.post("/v1/memory/conflicts/{conflict_id}/resolve")
async def resolve_conflict(conflict_id: str, request: Request):
    """裁决冲突：keep_old / use_new / manual。"""
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    body = await request.json()
    resolution = body.get("resolution", "")
    manual_value = body.get("manual_value")
    # 兼容前端命名：adopt_new → use_new
    if resolution == "adopt_new":
        resolution = "use_new"
    if resolution not in ("keep_old", "use_new", "manual"):
        return _bad_request("resolution must be one of: keep_old, use_new, manual")
    try:
        result = await svc.resolve_conflict(
            tenant_id=tenant_id,
            user_id=user_id,
            conflict_id=conflict_id,
            resolution=resolution,
            manual_value=manual_value,
        )
    except ValueError as e:
        return JSONResponse(status_code=404, content={"error": str(e)})
    return {"success": True, "conflict": result}


@router.delete("/v1/memory/conflicts/{conflict_id}")
async def delete_conflict(conflict_id: str, request: Request):
    """删除冲突（用户否认时调用）。"""
    svc = get_service()
    if svc is None:
        return _unavailable()
    tenant_id, user_id = _scope(request)
    if not user_id:
        return _bad_request("user_id is required")
    deleted = await svc.delete_conflict(tenant_id, user_id, conflict_id)
    if not deleted:
        return JSONResponse({"success": False, "error": "conflict not found"}, status_code=404)
    return {"success": True, "deleted": conflict_id}
